MAINTAINING.md

```diff
@@ -0,0 +1,147 @@
+2.8 Check for code efficiency
+
+
+## 3. - Test the PR
+Run tests to verify the correct behavior (see 1.5), or present **good** reasons
+for skipping some/all tests.
+
+
+## 4. - Review code style and documentation
+
+4.1 Check for sufficient high-level (module-level) documentation
+4.2 Verify function level documentation
+4.3 Are critical/hard to understand parts in the code documented?
+4.4 Check coding style (w.r.t. the coding conventions)
+4.5 Check grammar/spelling of documentation
+
+
+
+## Detailed Rationales, Explanations, and Examples
+
+WIP: extend this!
+
+### On 1.
+
+#### On 1.3
+- does this code properly maintain the Git history of previous contributions? (note by Martine: a PR can't change the Git history of the mainline repo => no force push possible, but maybe something else is meant here, if so, please specify)
+- does the boiler plate maintain copyrights of previous authors or contributors of ideas?
+
+### On 2.
+
+#### On 2.1
+- is the code just copied from somewhere in the RIOT code base?
+- does it make sense to join and factor out (parts of) the code with existing
+  code in RIOT?
+- is there needless code duplication withing the functions (e.g. same
+  instructions called multiple times)?
+- are internal functions used appropriately to reduce code duplication?
+
+#### On 2.2
+- large chunks of memory allocated on the stack?
+- needless over-provisioning of memory?
+- stack memory passed to different contexts?
+- Check all `size/length` parameters when passing memory:
+    - e.g. `sizeof(x)` vs. `strlen(x)`, especially `x` being a pointer...
+
+#### On 2.3
+- can all code paths be actually reached?
+    - e.g. always `true/false` conditions
+- are all critical sections (`lock/unlock`, `acquire/release`, ...) always
+  closed on every code path?
+
+#### On 2.4
+- are return values consistent with API documentation?
+
+#### On 2.5
+- run-time parameter checking vs. `assert()`
+- consistent usage of `DEBUG()` vs. `log_x(...)`
+
+#### On 2.6
+- are all module-internal variables and functions declared `static`?
+- are function variables moved into the smallest possible scope (cppcheck should
+  find this, right?)
+
+#### On 2.7
+- check for 'beginners' errors:
+    - logical not `!` vs bit inversion `~`
+    - logical vs. bit manipulation expressions (`x && y` vs `x & y`)
+- check signedness of variables
+- check variable width
+    - does target values fit into the selected type?
+    - `int/unsigned` is only 16-bit on msp430 and avr8...
+
+#### On 2.8
+- make sure we are not wasting any memory trough inefficient coding:
+```c
+typedef enum {
+    A,
+    B,
+    ...
+} foo_t;
+
+int bar(foo_t v)
+{
+    int abc;
+    ...
+
+    switch(v) {
+        case A:
+            abc = 23;
+            break;
+        case B:
+            abc = 42;
+            break;
+        ...
+    }
+    ...
+}
+
+/* VS */
+
+typedef enum {
+    A = 23,
+    B = 42,
+    ...
+} foo_t;
+
+int bar(foo_t v) {
+    int abc = v;
+    ...
+}
```

Comment:
I'm wondering if such concrete examples should rather go to the coding conventions or [best practices](https://github.com/RIOT-OS/RIOT/wiki/Best-Practice-for-RIOT-Programming)

Reply:
Yeah that sounds appropriate to me. That way, developers can also benefit from them. Perhaps all "detailed rationale" stuff could, rather than being here, be additions to our current coding conventions/best practices, for this reason? @haukepetersen any opinion?